Add an Easter event (impl #1662)

JapuDCret opened the pull request:
## What
closes #1662

Adds the SpecialEvent "Easter" and triggers the specific QuestLine "Togepi Egg Hunt".

## How

The QuestLine walks the Trainer through Veridian Forest, then Petalburg Woods and finally Ilex Forest, where afterwards the Trainer is rewarded with the holiday exclusive "Surprise Togepi".

<details>
  <summary>Currently used sprites (click to expand)</summary>
  
* Pokemon "Surprise Togepi"
  * my own creation, based on the base-sprite. I think it's ok, but not outstanding.
  ![175 2](https://user-images.githubusercontent.com/14893450/162493801-75406034-659a-4f8b-a4dd-9a621bd2a459.png)

* Trainer "Egg Hunter"
  * also my own quick creation, based on the sprite of "Collector". Someone more talented could definitely improve on that :D
  ![Egg Hunter](https://user-images.githubusercontent.com/14893450/162567810-74390f27-96dd-4b80-abd8-c0bc3eda8b35.png)
</details>

## Misc

Changes in `src/scripts/Game.ts` and the rework done in `src/scripts/specialEvents/SpecialEvents.ts` needed to be done, because otherwise the SpecialEvents code would be executed in the save-selection-screen and not in-game.

## Detailed walkthrough

<details>
  <summary>(click to expand)</summary>
  
* Start of the event and looking at Veridian Forest (1st location of Togepi)
  ![easer-event_issue1662_010_v2-kanto-before](https://user-images.githubusercontent.com/14893450/162567935-b418728f-1419-43cd-a1fd-8d3d4ce6f29b.png)
* Battle in Veridian Forest
  ![easer-event_issue1662_015_v2-kanto-during](https://user-images.githubusercontent.com/14893450/162567932-e58e4b5c-705d-41f2-8566-2d89a20a194d.png)
* Battle in Petalburg Woods
  ![easer-event_issue1662_025_v2-hoenn-during](https://user-images.githubusercontent.com/14893450/162567924-d2694a01-22cc-4f72-bdd1-5f070ee8f52e.png)
* Battle in Ilex Forest
  ![easer-event_issue1662_035_v2-johto-during](https://user-images.githubusercontent.com/14893450/162567901-f24e820e-6bf2-4e43-b055-0fc3cdacb44f.png)
* After successful completion of the Battle in Ilex Forest, the Trainer is rewarded the special "Surprise Togepi"!
  ![easer-event_issue1662_039_johto-after](https://user-images.githubusercontent.com/14893450/162493108-d9536ad3-b4aa-4f79-886e-1fffbde1bc24.png)
</details>

RedSparr0w commented on src/scripts/specialEvents/SpecialEvents.ts:
```diff
@@ -39,99 +39,114 @@ class SpecialEvents implements Feature {
-    }
-);
-
+document.addEventListener('pokeclicker.game.initialized', () => {
```
I think it would be better to move these to the `SpecialEvents.initialize()` function, so it's more consistent with how all the other `Feature` are handled.

JapuDCret replied:
thanks for the hint, works flawlessly, but required a small refactoring tho (since SpecialEvent.ts called `initialize()` in their constructor)

RedSparr0w commented on src/scripts/quests/QuestLineHelper.ts:
```diff
@@ -339,6 +339,63 @@ class QuestLineHelper {
+        findSurpriseTogepiForEasterQuestLine.addQuest(encounterTogepiInHoenn);
+
+        const togepiInJohtoSetup = () => {
+            dungeonList['Ilex Forest'].bossList.push(new DungeonTrainer('Easter basket', [ new GymPokemon('Surprise Togepi', 2700000, 100), ], { weight: 1, requirement: new GymBadgeRequirement(BadgeEnums.Elite_JohtoChampion) }));
```
Is there any reason we are using this particular location, or could it be changed?
As we already use this one for the `Grinch Celebi`.

JapuDCret replied:
I used Ilex Forest, because I was running with the theme, that in the parks/forests Egg hunts are happening and I can't think of a better suiting dungeon in Johto :/ Feedback appreciated

RedSparr0w commented on src/scripts/quests/QuestLineHelper.ts:
```diff
@@ -339,6 +339,63 @@ class QuestLineHelper {
+        const defeatTogepiInKanto = new CustomQuest(1, afterDefeatingTogepiInKanto, 'Erika reported, Togepi was last seen in Kanto. Go look for it!', App.game.statistics.pokemonDefeated[surpriseTogepi.id], 0, togepiInKantoSetup);
+        findSurpriseTogepiForEasterQuestLine.addQuest(defeatTogepiInKanto);
+
+        const togepiInHoennSetup = () => {
+            dungeonList['Petalburg Woods'].bossList.push(new DungeonTrainer('Easter basket', [ new GymPokemon('Surprise Togepi', 900000, 100), ], { weight: 1, requirement: new GymBadgeRequirement(BadgeEnums.Elite_HoennChampion) }));
+        };
+        const afterDefeatingTogepiInHoenn = () => {
+            Notifier.notify({
+                title: findSurpriseTogepiForEasterQuestLine.name,
+                message: 'I swear that was just a Togepi.. well maybe not.',
+                type: NotificationConstants.NotificationOption.info,
+                timeout: 3e4,
+            });
+
+            dungeonList['Petalburg Woods'].bossList = dungeonList['Petalburg Woods'].bossList.filter(boss => boss.name != 'Easter basket');
+        };
+        const encounterTogepiInHoenn = new CustomQuest(1, afterDefeatingTogepiInHoenn, 'Another report just came in, stating that they saw and strange egg on it\'s way to Hoenn!', App.game.statistics.pokemonDefeated[surpriseTogepi.id], 1, togepiInHoennSetup);
+        findSurpriseTogepiForEasterQuestLine.addQuest(encounterTogepiInHoenn);
```
Is there a reason we are using the order `Kanto → Hoenn → Johto`?

JapuDCret replied:
Yep, I didn't want it to be too linear - also finally encountering Togepi in Johto is also fitting, as it's native there.

RedSparr0w commented on src/scripts/Game.ts:
```diff
@@ -100,6 +100,8 @@ class Game {
 
         // If the player isn't on a route, they're in a town/dungeon
         this.gameState = player.route() ? GameConstants.GameState.fighting : GameConstants.GameState.town;
+
+        document.dispatchEvent(new CustomEvent('pokeclicker.game.initialized'));
```
As per below, use `SpecialEvents.initialize()` to be more consistent with other `Feature`.
```suggestion
```

RedSparr0w reviewed (changes requested):
Thanks for the PR!

Just a couple of small notes/changes.

Also, not a huge fan of the `Easter basket`, could we instead use an `Easter Bunny`/`Easter Bunnelby` or a trainer with bunny ears or something that more fits with our current sprite set?

JapuDCret commented:
Thanks for your feedback @RedSparr0w!

Here's what I changed
* Changed the `Easter basket` to an `Egg Hunter`
* Compressed the images
* Changed how the SpecialEvents are initialized (thanks for the hint)
* Fixed a typo and some ESLint findings

Your 2 remaining questions are up to preference, so please get back, if you want a change

RedSparr0w reviewed (approved):
Looks good, thanks for the PR! 👍 

I've swapped around Hoenn/Johto, just so newer players aren't waiting so long between each of the quest.

Nydaleclya commented:
Surprise Togepi is missing a shiny sprite.
